pytential/linalg/skeletonization.py

```diff
@@ -0,0 +1,790 @@
+
+    .. attribute:: source_farfield_block_builder
+    .. attribute:: target_farfield_block_builder
+
+        A callable that is used to evaluate farfield proxy interactions.
+        This should follow the calling convention of the constructor to
+        :class:`pytential.symbolic.matrix.FarFieldBlockBuilder`.
+
+    .. attribute:: nearfield_block_builder
+
+        A callable that is used to evaluate nearfield neighbour interactions.
+        This should follow the calling convention of the constructor to
+        :class:`pytential.symbolic.matrix.NearFieldBlockBuilder`.
+
+    .. automethod:: evaluate_source_farfield
+    .. automethod:: evaluate_target_farfield
+    .. automethod:: evaluate_source_nearfield
+    .. automethod:: evaluate_target_nearfield
+    """
+
+    exprs: np.ndarray
+    input_exprs: Tuple[sym.var, ...]
+    domains: Tuple[Hashable, ...]
+    context: Dict[str, Any]
+
+    # internal
+    weighted_sources: bool
+    weighted_targets: bool
+    source_farfield_block_builder: Callable[..., np.ndarray]
+    target_farfield_block_builder: Callable[..., np.ndarray]
+    nearfield_block_builder: Callable[..., np.ndarray]
+
+    def __post_init__(self):
+        assert len(self.input_exprs) == len(self.domains)
+
+    # {{{ internal
+
+    def _evaluate(self,
+            actx, places, cls, indices, expr, idomain,
+            context=None, **kwargs):
+        if context is None:
+            context = self.context
+
+        domain = self.domains[idomain]
+        dep_source = places.get_geometry(domain.geometry)
+        dep_discr = places.get_discretization(domain.geometry, domain.discr_stage)
+
+        return cls(actx,
+                dep_expr=self.input_exprs[idomain],
+                other_dep_exprs=(
+                    self.input_exprs[:idomain]
+                    + self.input_exprs[idomain+1:]),
+                dep_source=dep_source,
+                dep_discr=dep_discr,
+                places=places,
+                index_set=indices,
+                context=context,
+                **kwargs)(expr)
+
+    def _evaluate_nearfield(self, actx, places, indices, ibrow, ibcol, **kwargs):
+        expr = _prepare_nearfield_expr(
+                places, self.exprs[ibrow], auto_where=self.domains[ibcol])
+
+        cls = self.nearfield_block_builder
+        return self._evaluate(actx, places, cls, indices, expr, ibcol, **kwargs)
+
+    def _evaluate_farfield(self,
+            actx, places, cls, points, indices, ibrow, ibcol, *,
+            auto_where, qbx_forced_limit, **kwargs):
+        with _add_to_geometry_collection(places, points) as pxyplaces:
+            expr = _prepare_farfield_expr(
+                    pxyplaces, self.exprs[ibrow], auto_where=auto_where,
+                    qbx_forced_limit=qbx_forced_limit)
+
+            return self._evaluate(actx, places, cls, indices, expr, ibcol, **kwargs)
+
+    # }}}
+
+    # {{{ nearfield
```

Comment:
How much of this nearfield/farfield logic is essential to have?

Reply:
They mostly take care of the fact that some code is transposed for source vs target and some code uses proxy vs neighbor indices (the actual evaluation is obviously the same after those are taken care of). 

It could just have one function and use `partial` or something to wrap that stuff, but it didn't seem like it would be less (or clearer) code. If you still feel it could be cleaner on a second look I'll try to rewrite it :grin: